internal/proxy/index_params.go

```diff
@@ -0,0 +1,86 @@
+	IVFFLATParams = map[string]struct{}{common.IndexTypeKey: {}, common.MetricTypeKey: {}, common.DimKey: {}, indexparamcheck.NLIST: {}}
+	IVFSQ8Params  = map[string]struct{}{common.IndexTypeKey: {}, common.MetricTypeKey: {}, common.DimKey: {}, indexparamcheck.NLIST: {}}
+	IVFPQParams   = map[string]struct{}{common.IndexTypeKey: {}, common.MetricTypeKey: {}, common.DimKey: {}, indexparamcheck.NLIST: {},
+		indexparamcheck.IVFM: {}, indexparamcheck.NBITS: {}}
+	HNSWParams = map[string]struct{}{common.IndexTypeKey: {}, common.MetricTypeKey: {}, common.DimKey: {},
+		indexparamcheck.HNSWM: {}, indexparamcheck.EFConstruction: {}}
+	ANNOYParams = map[string]struct{}{common.IndexTypeKey: {}, common.MetricTypeKey: {}, common.DimKey: {},
+		indexparamcheck.NTREES: {}}
+	BINFLATParams    = map[string]struct{}{common.IndexTypeKey: {}, common.MetricTypeKey: {}, common.DimKey: {}}
+	BINIVFFLATParams = map[string]struct{}{common.IndexTypeKey: {}, common.MetricTypeKey: {}, common.DimKey: {}, indexparamcheck.NLIST: {}}
+	DISKANNParams    = map[string]struct{}{common.IndexTypeKey: {}, common.MetricTypeKey: {}, common.DimKey: {}, indexparams.MaxDegreeKey: {},
+		indexparams.SearchListSizeKey: {}, indexparams.PQCodeBudgetRatioKey: {}, indexparams.NumBuildThreadRatioKey: {},
+		indexparams.SearchCacheBudgetRatioKey: {}, indexparams.NumLoadThreadRatioKey: {}, indexparams.BeamWidthRatioKey: {}}
+	ScalarIndexParams = map[string]struct{}{common.IndexTypeKey: {}}
+)
+
+func checkIndexParams(indexParams map[string]string) error {
+	indexType, exist := indexParams[common.IndexTypeKey]
+	if !exist {
+		return fmt.Errorf("IndexType not specified")
+	}
+	switch indexType {
+	case indexparamcheck.IndexFaissIDMap:
+		return checkIndexParamsUtil(FLATParams, indexParams)
+	case indexparamcheck.IndexFaissIvfFlat:
+		return checkIndexParamsUtil(IVFFLATParams, indexParams)
+	case indexparamcheck.IndexFaissIvfSQ8:
+		return checkIndexParamsUtil(IVFSQ8Params, indexParams)
+	case indexparamcheck.IndexFaissIvfPQ:
+		return checkIndexParamsUtil(IVFPQParams, indexParams)
+	case indexparamcheck.IndexHNSW:
+		return checkIndexParamsUtil(HNSWParams, indexParams)
+	case indexparamcheck.IndexANNOY:
+		return checkIndexParamsUtil(ANNOYParams, indexParams)
+	case indexparamcheck.IndexFaissBinIDMap:
+		return checkIndexParamsUtil(BINFLATParams, indexParams)
+	case indexparamcheck.IndexFaissBinIvfFlat:
+		return checkIndexParamsUtil(BINIVFFLATParams, indexParams)
+	case indexparamcheck.IndexDISKANN:
+		return checkIndexParamsUtil(DISKANNParams, indexParams)
+	//case AUTOINDEX:
+	//	return checkIndexParamsUtil(AUTOINDEX, paramKey)
+	case DefaultStringIndexType, DefaultIndexType:
+		return checkIndexParamsUtil(ScalarIndexParams, indexParams)
+	default:
+		return fmt.Errorf("IndexType is not support: %s", indexType)
+	}
+}
+
+func checkIndexParamsUtil(expected map[string]struct{}, actual map[string]string) error {
+	for key := range actual {
+		if _, ok := expected[key]; !ok {
```

Comment:
So the logic of the check is to report an error if indexParams contains an unexpected key.
If indexParams does not include the expected key, is it not checked?

Reply:
We have additional logic `checkTrain` to check this.